Use `{cli}` for condition signalling

jamesmbaazam opened the pull request:
This PR partly addresses #542. It uses the use of base R to various {cli} functionalities for signaling messages, warnings, errors, and strings. 

Tests have been updated to reflect some of the changes introduced by {cli} formatting.

This PR fixes #673

seabbs commented:
How are we getting along with this @jamesmbaazam? I think @nikosbosse and co all think its a good idea to make use of `{cli}` so very keen to move this forward if we can.

jamesmbaazam commented:
Thanks for the ping @seabbs. I apologize for the delay in pushing it forward. 

The switch from `base` to `cli` is straightforward as you know, but the more important issue I've been grappling with is ensuring that the use of `cli` improves debugging. Currently, I'm unable to add the function call to warning signals.  In general [warnings are quite difficult to handle in R](https://adv-r.hadley.nz/conditions.html#warnings).

I'm inclined to suggest we consider changing some of the warnings here to errors to improve debugging. For example, `run_safely()` converts errors to warnings, which I guess is meant to improve user experience but I think it comes at the cost of difficulty to debug. Other examples include throwing a warning in `ae_median_quantile()` and returning NAs while an informative error could help the user fix the issue.

What do you think?

nikosbosse commented:
@jamesmbaazam apologies for the delay in responding - apparently I failed to click on "comment" the last time I wrote a reply... 

Maybe the underlying problem really is that the default scoring rules are currently a bit hard to work with. We're providing the default scoring rules e.g. in `rules_quantile()` and removing several of those or just selecting several of those is slightly annoying. And if one of your scoring rules errors then the call to `score()` errors. Maybe thinking a bit more about the default scoring rules / adding something that checks whether they are applicable at all would be helpful. For example, 90% interval coverage is part of the default scoring rules, but that can of course only be computed if the interval is actually there. 

That aside, I'm not in principle opposed to dropping the conversion from warnings to errors. @seabbs what do you think?

seabbs commented:
> That aside, I'm not in principle opposed to dropping the conversion from warnings to errors. @seabbs what do you think?

1. We should promote this discussion to its own issue for discussion.
2. If we were to do that then the consequence would be `score` fast failing if users have scoring metrics that don't work. Historically @nikosbosse was particularly opposed to that and to be honest I can see why.
3. So it comes down to does localising the error outweigh making `score` robust to poorly set scoring rules?

nikosbosse commented:
Added an issue here: https://github.com/epiforecasts/scoringutils/issues/646 Agree we should discuss this separately from the switch to `cli`

seabbs commented:
How are we getting on with this? It would be amazing to be able to merge in soonish so that 1. It doesn't become really horrible keeping this up to date and 2. We can do whatever additional iteration we want to do on the cli ahead of the new big release.

nikosbosse commented:
I went ahead and fixed the merge conflicts

nikosbosse commented:
@jamesmbaazam I saw that you're not yet a contributor on the DESCRIPTION file. Could you add yourself, please? Thank you very much!

nikosbosse commented on R/check-input-helpers.R:
```diff
@@ -210,20 +233,20 @@ check_columns_present <- function(data, columns) {
-      "Columns '", paste(missing, collapse = "', '"), "' not found in data"
+  if (length(missing) >= 1) {
+    return(
+      cli_inform(
```
I think this should not return `cli_inform`. As is, the function immediately outputs a message to the console. Previously, the function returned a string with a message that would then be handled by the function calling `check_columns_present()`. 

With the current updated behaviour, you always get a message when calling `as_forecast()` about either `sample_id` or `quantile_level` missing.

nikosbosse replied:
I think the latest version is the correct version!
This is marked as outdated and I think if you click on it you get to the original version I made a comment for.

nikosbosse commented:
> Some of the functions I worked on are actually now deleted 😭

Ayyy apologies! Here is a picture of a friendly kitten to keep up developer morale! 

<img width = 300, src="https://i.pinimg.com/736x/20/ba/e2/20bae2b032f4b4109e69e835eaad158a.jpg">

Also let me know if there is anything I can help with, happy to jump on.

jamesmbaazam commented:
> I went ahead and fixed the merge conflicts and also added `cli` to the imports in the DESCRIPTION. Checks are currently failing, but I _think_ that's not due to my fixing the merge conflicts. If so I'm happy to clean this up (and am happy to work on the PR in general if you think it would be useful, @jamesmbaazam)

Thanks, Nikos. I'll push what I have locally tonight then we can see what's left to be done.

nikosbosse commented:
Also @jamesmbaazam if you like you can also just push anything you have to a different branch, make a PR to this one and I'll fix the merge conflicts. It might be easier for me to do that than for you because I authored all the PRs that led to the merge conflicts...

jamesmbaazam commented:
This is as far as I can get tonight. The checks are expected to fail because of some duplicated rows in the dataset used for the tests. Will check in detail tomorrow.

nikosbosse commented:
Nice! Thank you for your heroic efforts!

<img src="https://img.freepik.com/premium-photo/cat-superhero-costume-with-superman-costume_36682-642.jpg" width=400>

<img src="https://static.printler.com/cache/d/5/d/8/e/2/d5d8e20bf1e33115bdc0df13dc726852295bbc53.jpg" width=400>

jamesmbaazam commented:
New wallpaper loading...

seabbs commented:
The image based inspiration is amazing. Well done both on this. It seems like a nightmare but also one that is very worth having I think

jamesmbaazam commented:
> The image based inspiration is amazing. 

So creative, I'm stealing it.

jamesmbaazam commented:
@nikosbosse The current failures are as a result of the `example-binary` dataset getting overwritten at some point during this PR, so I'm going to recreate it using the script in `./inst/...`. I confirmed in the `main` branch that this failure wasn't happening. The dataset on this branch has duplicate rows where as that on `main` doesn't.

nikosbosse commented:
@jamesmbaazam maybe it's easier to just copy over the file from the other branch to replace the one you already have? The script forces you to download parts of the European COVID-19 forecast hub repo which may be cumbersome (although independently verifying that script might be of course be worthwhile). 
Let me know if I can help with anything. 
Also should I go ahead and deal with the merge conflicts again?

jamesmbaazam commented:
> @jamesmbaazam maybe it's easier to just copy over the file from the other branch to replace the one you already have? The script forces you to download parts of the European COVID-19 forecast hub repo which may be cumbersome (although independently verifying that script might of course be worthwhile). Let me know if I can help with anything. 

The errors here will persist unless we push the correct data onto this branch but I think it is currently being ignored.

> Also should I go ahead and deal with the merge conflicts again?

Yes, please go ahead. Thanks. If you can fix the errors too, that would be great. I'll also open it up for review now.

nikosbosse commented:
I think the source of the error was that one of the tests deleted the model column of `example_binary`. So the actual package file was fine, it was just that the test altered it (that's why updating the data didn't help). Should be fixed now. I'll have a look at the PR now.

nikosbosse commented on .gitignore:
```diff
@@ -19,3 +19,5 @@ docs
 .DS_Store
 .vscode
 README.html
+/doc/
```
just out of curiosity, what do `/doc/` and `/Meta/` do?

jamesmbaazam replied:
I often build individual aspects of the package before running the full `devtools::check()`. These files were generated from a `devtools::build_vignettes()` run. That's when they were created. From the helpfile (`?devtools::build_vignettes`):


"Builds package vignettes using the same algorithm that ⁠R CMD build⁠ does. This means including non-Sweave vignettes, using makefiles (if present), and copying over extra files. The files are copied in the 'doc' directory and an vignette index is created in 'Meta/vignette.rds', as they would be in a built package. 'doc' and 'Meta' are added to .Rbuildignore, so will not be included in the built package. These files can be checked into version control, so they can be viewed with browseVignettes() and vignette() if the package has been loaded with load_all() without needing to re-build them locally."

nikosbosse commented on R/check-input-helpers.R:
```diff
@@ -274,7 +293,7 @@ check_data_columns <- function(data) {
   if (problem) {
     return(
-      "Found columns `quantile_level` and `sample_id`. Only one of these is allowed"
+      "Found columns `quantile` and `sample_id`. Only one of these is allowed"
```
```suggestion
      "Found columns `quantile_level` and `sample_id`. Only one of these is allowed"
```

nikosbosse commented on R/metrics-quantile.R:
```diff
@@ -253,11 +255,16 @@ interval_coverage <- function(observed, predicted,
+    cli_warn(
+      c(
+        "!" = "To compute the interval coverage for an interval range of
+        {interval_range}, the {.val {necessary_quantiles}} quantiles
```
```suggestion
        {interval_range}%, the {.val {necessary_quantiles}} quantiles
```

the old version had a %-sign there. is it possible to get that back in like that?

nikosbosse commented on R/metrics-quantile.R:
```diff
@@ -310,6 +317,8 @@ interval_coverage <- function(observed, predicted,
 #' The interval coverage deviation is then averaged across all prediction
 #' intervals. The median is ignored when computing coverage deviation.
 #' @inheritParams wis
+#' @importFrom rlang caller_env
```
is this used in the function?

jamesmbaazam replied:
Nope, my bad. Was left over from initial attempts to include the function call in cli_warn(). I've removed all the `rlang` imports and the package itself from the `Imports` field because it is no longer used.

nikosbosse commented on R/get_-functions.R:
```diff
@@ -108,7 +115,7 @@ get_type <- function(x) {
   assert_numeric(as.vector(x))
   if (all(is.na(as.vector(x)))) {
-    stop("Can't get type: all values of are NA")
+    cli_abort("Can't get type: all values of are NA")
```
```suggestion
    cli_abort("Can't get type: all values of are {.val {NA}}.")
```

jamesmbaazam replied:
Is it supposed to read "all values ~of~ are NA"? There's something missing currently.

nikosbosse commented on R/metrics-quantile.R:
```diff
@@ -532,10 +560,15 @@ bias_quantile_single_vector <- function(observed, predicted,
+      c(
+        "x" = "In order to compute the absolute error of the median,
+        {.var 0.5} must be among the quantiles given.",
+        "i" = "Returning `NA`."
```
```suggestion
        "i" = "Returning {.val {NA}}."
```

nikosbosse commented on R/metrics-quantile.R:
```diff
@@ -532,10 +560,15 @@ bias_quantile_single_vector <- function(observed, predicted,
+    cli_warn(
+      c(
+        "x" = "In order to compute the absolute error of the median,
+        {.var 0.5} must be among the quantiles given.",
```
is there a difference between `.var` and `.val`? Should this be `.val` here?

jamesmbaazam replied:
You're right. It should be `.val` for values and `.var` for variables.

nikosbosse commented on R/pairwise-comparisons.R:
```diff
@@ -94,34 +101,49 @@ pairwise_comparison <- function(
+      #nolint start: keyword_quote_linter object_usage_linter
+      cli_warn(
+        c(
+          "!" = "After removing NA values for {.var {metric}},
```
```suggestion
          "!" = "After removing {.val {NA}} values for {.var {metric}},
```

nikosbosse commented on R/pairwise-comparisons.R:
```diff
@@ -132,16 +154,24 @@ pairwise_comparison <- function(
+      cli_inform(
+        c(
+          "!" = "relative skill can only be computed if `by` is different from the
+        unit of a single forecast.",
```
I still don't really understand how linebreaks work in `cli`. Should this be a linebreak? Is it even easily possible to wrap code in the 80char line limit?

jamesmbaazam replied:
As far as I know, it concatenates the string and removes all `\n` values, so basically, it prints a single string wrapped in the console. To get new lines, you need to use bullets, paragraphs, or texts (`cli_text`).

> Should this be a linebreak? Is it even easily possible to wrap code in the 80char line limit?

The line break here is just visual and to escape the 80-character limit but it is printed as a single string.

jamesmbaazam replied:
Yes, exactly.

nikosbosse commented on R/pairwise-comparisons.R:
```diff
@@ -338,7 +374,9 @@ compare_two_models <- function(scores,
   if (!("model" %in% names(scores))) {
-    stop("pairwise comparisons require a column called 'model'")
+    cli_abort(
+      "pairwise comparisons require a column called 'model'"
```
```suggestion
      "pairwise comparisons require a column called {.val {"model"}}."
```
would that work?

jamesmbaazam replied:
I think we want to use `.var` here for variable?

nikosbosse commented on R/pit.R:
```diff
@@ -61,6 +61,7 @@
 #' `length(observed) * n_replicates`
 #' @seealso [pit()]
 #' @importFrom stats runif
+#' @importFrom cli cli_inform
```
```suggestion
#' @importFrom cli cli_abort cli_inform
```

nikosbosse commented on R/utils.R:
```diff
@@ -161,22 +163,44 @@ print.forecast_binary <- function(x, ...) {
-    cat("Forecast type:\n")
-    print(forecast_type)
+    cli_text(
+      col_blue(
```
Love this! :)

nikosbosse commented on tests/testthat/test-score.R:
```diff
@@ -108,7 +108,7 @@ test_that(
     }
 
-    df <- example_binary[model == "EuroCOVIDhub-ensemble" &
+    df <- copy(example_binary)[model == "EuroCOVIDhub-ensemble" &
```
```suggestion
    df <- data.table::copy(example_binary)[model == "EuroCOVIDhub-ensemble" &
```

(trying to use `::` in tests because at some point we want to remove the call to `library(data.table)` in the setup script to not accidentally mask any missing functionality.

nikosbosse reviewed (changes requested):
Thanks so much, @jamesmbaazam. This looks really good and was a herculean task. I only made a few minor comments and otherwise think this is good to go. 

<img src="https://external-preview.redd.it/T6P4lga9Ce3ZGiat9E5QFNaXvfP7DslLQjf1N68KW-8.jpg?auto=webp&s=efe097df4053f1897c91c93cab9faf00385d000c" width=500>

nikosbosse commented:
<img src="https://external-preview.redd.it/T6P4lga9Ce3ZGiat9E5QFNaXvfP7DslLQjf1N68KW-8.jpg?auto=webp&s=efe097df4053f1897c91c93cab9faf00385d000c" width=500>

jamesmbaazam commented on NEWS.md:
```diff
@@ -44,6 +44,8 @@ The update introduces breaking changes. If you want to keep using the older vers
 - Added unit tests for `interval_coverage_quantile()` and `interval_coverage_dev_quantile()` in order to make sure that the functions provide the correct warnings when insufficient quantiles are provided.
 - Documentation pkgdown pages are now created both for the stable and dev versions.
+- Switched to using `cli` for condition handling and signalling.
+- Added tests for all the `check_*()` and `test_*()` functions. See #583 by @jamesmbaazam and reviewed by <insert reviewers>.
```
```suggestion
- Added tests for all the `check_*()` and `test_*()` functions. See #583 by @jamesmbaazam and reviewed by @nikosbosse and @seabbs.
```

jamesmbaazam commented on R/metrics-quantile.R:
```diff
@@ -532,10 +560,15 @@ bias_quantile_single_vector <- function(observed, predicted,
+    cli_warn(
+      c(
+        "x" = "In order to compute the absolute error of the median,
+        {.var 0.5} must be among the quantiles given.",
```
```suggestion
        {.val 0.5} must be among the quantiles given.",
```

jamesmbaazam commented on R/pairwise-comparisons.R:
```diff
@@ -338,7 +374,9 @@ compare_two_models <- function(scores,
   if (!("model" %in% names(scores))) {
-    stop("pairwise comparisons require a column called 'model'")
+    cli_abort(
+      "pairwise comparisons require a column called 'model'"
```
```suggestion
      "pairwise comparisons require a column called {.val model}"
```

nikosbosse replied:
```suggestion
      "!" = "pairwise comparisons require a column called 'model'"
```

jamesmbaazam commented on R/pairwise-comparisons.R:
```diff
@@ -132,16 +154,24 @@ pairwise_comparison <- function(
+      cli_inform(
+        c(
+          "!" = "relative skill can only be computed if `by` is different from the
+        unit of a single forecast.",
```
```suggestion
           unit of a single forecast.",
```

nikosbosse commented on R/pairwise-comparisons.R:
```diff
@@ -188,7 +219,9 @@ pairwise_comparison_one_group <- function(scores,
   if (!("model" %in% names(scores))) {
-    stop("pairwise compairons require a column called 'model'")
+    cli_abort(
+      "pairwise compairons require a column called 'model'"
```
```suggestion
      "!" = "pairwise compairons require a column called 'model'"
```

There are a few `cli` statements which are missing a `"!" =`
I'll make some suggested edits when I see them, but otherwise we could also just live that to a future PR

jamesmbaazam replied:
Thanks. `cli_abort()` adds it by default, so where we ain't providing extra information, I take advantage of that to avoid the trouble of having to concatenate the bullet (in other words, I'm lazy).

nikosbosse commented on R/pairwise-comparisons.R:
```diff
@@ -267,7 +300,9 @@ pairwise_comparison_one_group <- function(scores,
     if (length(baseline_theta) == 0) {
-      stop("Baseline model ", baseline, " missing.")
+      cli_abort(
+        "Baseline model {.var {baseline}} missing."
```
```suggestion
        "!" = "Baseline model {.var {baseline}} missing."
```

nikosbosse commented on R/pairwise-comparisons.R:
```diff
@@ -405,10 +443,15 @@ infer_rel_skill_metric <- function(scores) {
+    #nolint start: keyword_quote_linter
+    cli_abort(
+      c(
+        "Automatically assigning a metric to compute relative skills
```
```suggestion
        "!" = "Automatically assigning a metric to compute relative skills
```

(I think that function will be dropped anyway soon see #673

nikosbosse commented on R/pit.R:
```diff
@@ -93,15 +94,17 @@ pit_sample <- function(observed,
   if (missing("observed") || missing("predicted")) {
-    stop("`observed` or `predicted` missing in function 'pit_sample()'")
+    cli_abort(
+      "`observed` or `predicted` missing in function {.fn pit_sample}."
```
```suggestion
      "!" = "`observed` or `predicted` missing in function {.fn pit_sample}."
```

nikosbosse commented on R/pit.R:
```diff
@@ -93,15 +94,17 @@ pit_sample <- function(observed,
-    message(
-      "you need more than one observation to assess uniformity of the PIT"
+    cli_inform(
+      "You need more than one observation to assess uniformity of the PIT."
```
```suggestion
      "i" = "You need more than one observation to assess uniformity of the PIT."
```

nikosbosse commented on R/pit.R:
```diff
@@ -111,18 +114,16 @@ pit_sample <- function(observed,
-      "'predicted' should be a matrix. Instead `%s` was found",
-      class(predicted)[1]
+    cli_abort(
+      "`predicted` should be a {.cls matrix},
```
```suggestion
      "!" = "`predicted` should be a {.cls matrix},
```

nikosbosse commented on R/pit.R:
```diff
@@ -111,18 +114,16 @@ pit_sample <- function(observed,
-      "Mismatch: 'observed' has length `%s`, but 'predicted' has `%s` rows.",
-      n, nrow(predicted)
+    cli_abort(
+      "Mismatch: 'observed' has length {.var {n}},
```
```suggestion
      "!" = "Mismatch: 'observed' has length {.var {n}},
```

nikosbosse commented on R/utils.R:
```diff
@@ -97,11 +98,13 @@ run_safely <- function(..., fun) {
-      "Function execution failed, returning NULL. Error: ",
-      msg
+    cli_warn(
+      "Function execution failed, returning NULL.
```
```suggestion
      "!" = "Function execution failed, returning NULL.
```

nikosbosse commented on R/utils.R:
```diff
@@ -144,10 +148,8 @@ print.forecast_binary <- function(x, ...) {
-      "Error in validating forecast object:\n",
-      validation_msg
+    cli_warn(
+      "Error in validating forecast object: {validation}."
```
```suggestion
      "!" = "Error in validating forecast object: {validation}."
```

nikosbosse commented on R/utils.R:
```diff
@@ -161,22 +163,44 @@ print.forecast_binary <- function(x, ...) {
   if (inherits(forecast_type, "try-error")) {
-    message("Could not determine forecast type due to error in validation.")
+    cli_inform(
+      "Could not determine forecast type due to error in validation."
```
```suggestion
      "!" = "Could not determine forecast type due to error in validation."
```

nikosbosse commented on R/utils.R:
```diff
@@ -161,22 +163,44 @@ print.forecast_binary <- function(x, ...) {
   if (length(forecast_unit) == 0) {
-    message("Could not determine forecast unit")
+    cli_inform(
+      "Could not determine forecast unit."
```
```suggestion
      "!" = "Could not determine forecast unit."
```

nikosbosse reviewed (commented):
I just added a few `"!" = "` statements here and there. Not sure they are all helpful, so please feel free to either accept or reject them as you see fit!

nikosbosse commented:
@jamesmbaazam do you think this is good to go? 🚀

jamesmbaazam commented:
> @jamesmbaazam do you think this is good to go? 🚀

Are we gonna ignore the codecov error? 😆

nikosbosse commented:
> Are we gonna ignore the codecov error? 😆

I would of course be very happy for the code coverage holes to be fixed. But since this has already been a really long and cumbersome PR and since the holes in the coverage were pre-existent, I wouldn't want to _force_ you/us to fix them in this PR.

nikosbosse commented:
Recent changes: 
- started updating tests
- deleted the following input check from `assert_dims_ok_point()` because I think the check can never be reached due to checks above: 
```

dim_p <- dim(predicted)
--
246 |   | if (!is.null(dim_p) && (length(dim_p) > 1) && (dim_p[2] > 1)) {
247 |   | #nolint start: keyword_quote_linter object_usage_linter
248 |   | cli_abort(
249 |   | c(
250 |   | "!" = "`predicted` must be a vector or a matrix with one column.",
251 |   | "i" = "Found {dim(predicted)[2]} columns."
252 |   | )
253 |   | )
254 |   | #nolint end
255 |   | }
```

nikosbosse commented:
I added a bunch of tests. Codecov is still failing and I'm not entirely sure why. Some of the lines it complains about are definitely covered by tests. I think my solution would mostly be ignoring it 🤷 

@jamesmbaazam when I ran the tests again an unused snapshot file came up (last commit) that I deleted. Is that a file that was previously used? 

Another (minor) thing I noticed: when running tests, the `test-utils.R` file now produces a bunch of printed output: 
<img width="975" alt="image" src="https://github.com/epiforecasts/scoringutils/assets/37978797/22adfa14-25e1-49ae-9231-e37ee77fa4ad">

`cli_text()` can apparently be suppressed with `suppressMessages()`. Maybe it makes sense to wrap the print statements in `suppressMessages()`? Not a big thing in any case. Just noting it.

nikosbosse commented:
@seabbs I think the PR is good to go, but maybe you would like to have a quick look at it

jamesmbaazam commented:
> Also I can't merge it without git getting angry at me for bypassing branch protection rules

Technically, you're an accomplice now so Sam might have to take a look.

The only thing I'm not satisfied with is the fact that the texts are printed in white. I would prefer a different color to differentiate between signalled conditions and printed output. What do you think?

nikosbosse commented:
> The only thing I'm not satisfied with is the fact that the texts are printed in white. 

I like the different colours. However, since it's not a breaking change maybe we should do it in a separate PR?

jamesmbaazam commented:
> since it's not a breaking change maybe we should do it in a separate PR?

Yep, definitely. Let's see how the current goes and improve it in the future. 

Also, I'll respond to your other comments tomorrow. Stepping away from my PC for tonight.

jamesmbaazam commented:
> `cli_text()` can apparently be suppressed with `suppressMessages()`. Maybe it makes sense to wrap the print statements in `suppressMessages()`? Not a big thing in any case. Just noting it.

Yes, they can be suppressed but I'm often wary of suppressing messages and warnings especially in tests unless we're sure we're suppressing the exact messages that we want to suppress (that's a lot of suppressions in one sentence).

jamesmbaazam commented:
> @jamesmbaazam when I ran the tests again an unused snapshot file came up that I deleted ([3028945](https://github.com/epiforecasts/scoringutils/pull/583/commits/30289456a8a52207bdd7c0778786edbcf2b927ad)). Is that a file that was previously used?

Is this related to me changing an existing test to a snapshot test here [`bb2ddd9` (#583)](https://github.com/epiforecasts/scoringutils/pull/583/commits/bb2ddd97de8e8115599b28643ec852db9a039842)?

seabbs commented:
> Yes, they can be suppressed but I'm often wary of suppressing messages and warnings especially in tests unless we're sure we're suppressing the right messages (that's a lot of suppressions in one sentence).

I really don't like having noisy tests as it makes it hard to look over them quickly. 

For this and other small discussion points and issues can wee have a few issues prior to merging this?

seabbs commented on R/check-inputs-scoring-functions.R:
```diff
@@ -123,22 +123,27 @@ assert_input_interval <- function(observed, lower, upper, interval_range) {
-    stop(
-      "All values in `upper` need to be greater than or equal to ",
-      "the corresponding values in `lower`"
+    cli_abort(
```
This lack of coverage looks real to me. I think we should promote fixing this to an issue vs dealing with here.

nikosbosse replied:
Unsure 🤷 
I specifically added this test: 
https://github.com/epiforecasts/scoringutils/blob/72d9090309d3fea37b396e25c2637d3b2521cbf1/tests/testthat/check-inputs-scoring-functions.R#L13
<img width="795" alt="image" src="https://github.com/epiforecasts/scoringutils/assets/37978797/69158a23-9634-42f8-8568-e1cbcbbf812f">
But I added an issue (https://github.com/epiforecasts/scoringutils/issues/686)

seabbs commented on R/utils.R:
```diff
@@ -157,26 +145,38 @@ print.forecast_binary <- function(x, ...) {
+    )
   }
 
   cat("\n")
```
Do we still need this?

nikosbosse replied:
+ should we discuss this in #621?

seabbs reviewed (changes requested):
This all looks good to me and is a big improvement over what we had! As above I think the remaining things to do here are to 1. Resolve conflicts, 2. open some new issues.

nikosbosse commented:
I resolved the merge conlict and added the following issues: 
- https://github.com/epiforecasts/scoringutils/issues/685
- https://github.com/epiforecasts/scoringutils/issues/686

And there is an existing one: 
- #621 

@seabbs does that address everything we should do before merging?

nikosbosse commented:
woohoo! Thank you @jamesmbaazam and @seabbs!

<img src="https://www.rd.com/wp-content/uploads/2018/02/28_Adorable-Puppy-Pictures-that-Will-Make-You-Melt_529323343_Sigma_S.jpg?fit=700,700" width = 400>

<img src="https://media.istockphoto.com/id/636475496/photo/portrait-of-brown-puppy-with-bokeh-background.jpg?s=612x612&w=0&k=20&c=Ot63dQOYplm0kLJdlSVWbtKGwGkuZfnfdwH5ry9a6EQ=" width = 400>

seabbs commented:
Shotgun being the second one
